modules/discretedata/dataset.h

```diff
@@ -0,0 +1,209 @@
+#include "channels/bufferchannel.h"
+#include "connectivity/connectivity.h"
+
+namespace inviwo {
+namespace dd {
+
+typedef std::shared_ptr<Channel> SharedChannel;
+typedef std::shared_ptr<const Channel> SharedConstChannel;
+
+struct ChannelCompare {
+    bool operator()(const std::pair<std::string, GridPrimitive>& u,
+                    const std::pair<std::string, GridPrimitive>& v) const {
+        return (u.second < v.second) || u.first.compare(v.first) < 0;
+    }
+};
+// Map used for storing and querying channels by name and GridPrimitive type.
+typedef std::map<
+    std::pair<std::string, GridPrimitive>  // Hashing by both name and GridPrimitive type
+    ,
+    SharedConstChannel  // Shared channels, type information only as meta property
+    ,
+    ChannelCompare>  // Lesser operator on string-Primitve pairs
+    BaseChannelMap;
+
+class IVW_MODULE_DISCRETEDATA_API DataChannelMap {
+public:
+    DataChannelMap() = default;
+    virtual ~DataChannelMap() = default;
+
+    /** Default copy shares Channels */
+    DataChannelMap(const DataChannelMap& copy) = default;
+
+    // Methods
+public:
+    /** Add a new channel to the set
+    *   @param channel Pointer to data, takes memory ownership
+    *   @return Shared pointer for further handling
+    */
+    SharedChannel addChannel(Channel* channel);
+
+    /** Add a new channel to the set
+    *   @param channel Shared pointer to data, remains valid
+    */
+    void addChannel(SharedConstChannel channel);
+
+    /** Returns the first channel from an unordered list.
+    */
+    SharedConstChannel getFirstChannel() const;
+
+    /** Returns the first channel from an unordered list.
+    */
+    template <typename T>
+    std::shared_ptr<const DataChannel<T>> getFirstChannel() const;
+
+    /** Returns the specified channel, returns first instance found
+    *   @param name Unique name of requested channel
+    *   @param definedOn GridPrimitive type the channel is defined on, default 0D vertices
+    */
+    SharedConstChannel getChannel(const std::string& name,
+                                  GridPrimitive definedOn = GridPrimitive::Vertex) const;
+
+    /** Returns the specified channel if it is in the desired format, returns first instance found
+    *   @param name Unique name of requested channel
+    *   @param definedOn GridPrimitive type the channel is defined on, default 0D vertices
+    */
+    template <typename T>
+    std::shared_ptr<const DataChannel<T>> getChannel(
+        const std::string& name, GridPrimitive definedOn = GridPrimitive::Vertex) const;
+
+    /** Returns the specified channel if it is in the desired format, returns first instance found
+    *   @param name Unique name of requested channel
+    *   @param definedOn GridPrimitive type the channel is defined on, default 0D vertices
+    */
+    template <typename T>
+    std::shared_ptr<const DataChannel<T>> getChannel(
+        const char* name, GridPrimitive definedOn = GridPrimitive::Vertex) const;
```

Comment:
don't think we need overloads for const char*

Reply:
I like to be able to just throw in a "name" without explicitly generating a std::string.

You talked about these basic_string_views, I'll look into it.

Reply:
Don't think you need to worry about that performance impact. Would say that is premature optimisation. But you still find it important, then from the api point of view you should probably have that overload for all of the functions that takes string arguments, like the non template version and getAsBuffer...
About std::string_view since it is c++17 we don't use it in core so far. since we only require c++14. But since this is a completely new module and not a mandatory one for running inviwo... I think we could require c++17 in here if you want to. Would be kind of a waste not using the lastest standard when doing this from scratch...

Reply:
I remember there was some reason. Compiling it with actually working code right now.

Reply:
Actually, the error was the missing `const`. Works fine now. Removed.